package BinarySearch;

public class hSearchRotatedArray {

    public static int search(int[] arr, int target) {

        int low = 0, high = arr.length - 1;

        while (low <= high) {

            int mid = low + (high - low) / 2;

            // If found
            if (arr[mid] == target) {
                return mid;
            }

            // 🔹 Check if LEFT half is sorted
            if (arr[low] <= arr[mid]) {

                // target lies in left sorted half
                if (arr[low] <= target && target < arr[mid]) {
                    high = mid - 1;
                } 
                else {
                    low = mid + 1;
                }
            }

            // 🔹 Right half is sorted
            else {

                if (arr[mid] < target && target <= arr[high]) {
                    low = mid + 1;
                } 
                else {
                    high = mid - 1;
                }
            }
        }

        return -1;
    }

    public static void main(String[] args) {
        int[] arr = {4, 5, 6, 7, 1, 2, 3};

        System.out.println(search(arr, 6)); // Output: 2
        System.out.println(search(arr, 2)); // Output: 5
    }
}